### PR TITLE
Worker Abstraction Update

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
Upgrades the FPS Starter Project with the changes made in https://github.com/spatialos/gdk-for-unity/pull/981. 

Also did quite a lot of driveby changes on the sim player flow.

Commits are distinct, helpful for reviewing the PR 

#### Tests

- [x] Local & PC
- [x] Local & Mobile
- [x] Cloud & PC
- [x] Cloud & Mobile
- [x] Session Flow & PC
- [x] Session Flow & Mobile

#### Left todo

- [x] Changelog
- [x] Update `gdk.pinned` once https://github.com/spatialos/gdk-for-unity/pull/981 is merged

### CODE REVIEW

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs [hunk 1]
```diff
@@ -1,209 +1,339 @@
...
 
-    private readonly List<GameObject> simulatedPlayerConnectors = new List<GameObject>();
+        public GameObject SimulatedPlayerWorkerConnector;
 
-    protected override async void Start()
-    {
-        Application.targetFrameRate = 60;
-        await AttemptConnect();
-    }
+        // TODO: Nice Unity tooltips
```
Comment (zeroZshadow) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: TODO

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs [hunk 2]
```diff
@@ -1,209 +1,339 @@
...
-                StartCoroutine(MonitorSimulatedPlayers());
+                connectPlayersWithDevAuth = deploymentName != "local";
             }
+            else
+            {
+                // We are probably in the Editor.
+                connectPlayersWithDevAuth = false;
+            }
+
+            Application.targetFrameRate = 60;
```
Comment (zeroZshadow) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: Open a ticket for making these configurable

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs [hunk 3]
```diff
@@ -1,209 +1,339 @@
...
+        {
+            var builder = new SpatialOSConnectionHandlerBuilder()
+                .SetConnectionParameters(CreateConnectionParameters(WorkerUtils.SimulatedPlayerCoordinator));
 
-    private IEnumerator MonitorSimulatedPlayers()
-    {
-        while (Worker.Connection?.GetConnectionStatusCode() == ConnectionStatusCode.Success)
+            if (Application.isEditor)
+            {
+                builder.SetConnectionFlow(new ReceptionistFlow(CreateNewWorkerId(WorkerUtils.SimulatedPlayerCoordinator)));
```
Comment (zeroZshadow) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: More workerID outside of `if`

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs [hunk 4]
```diff
@@ -1,209 +1,339 @@
...
             {
-                DefaultSimulatedPlayerCount = newCount;
-                return true;
+                if (token.IsCancellationRequested)
+                {
+                    throw new TaskCanceledException();
+                }
+
+                Worker.LogDispatcher.HandleLog(LogType.Log, new LogEvent("Waiting for required worker flags.."));
+                await Task.Delay(TimeSpan.FromSeconds(2), token);
```
Comment (zeroZshadow) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: Magic timeout

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs [hunk 5]
```diff
@@ -1,209 +1,339 @@
...
+            var worldSize = GetWorldSize();
+            return new Bounds(Worker.Origin, MapBuilder.GetWorldDimensions(worldSize));
         }
-    }
 
-    public Bounds GetWorldBounds()
-    {
-        var worldSize = GetWorldSize();
-        return new Bounds(Worker.Origin, MapBuilder.GetWorldDimensions(worldSize));
+        private class WorkerFlagTracker
```
Comment (zeroZshadow) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: Should be disposable with an unregister

File: workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs (hunk 2 above)
Comment (jamiebrynes7) [reply workers/unity/Assets/Fps/Scripts/SetupLogic/SimulatedPlayerCoordinatorWorkerConnector.cs]: https://improbableio.atlassian.net/browse/UTY-2115

File: workers/unity/Assets/Fps/Scripts/SetupLogic/WorkerConnectorBase.cs [hunk 6]
```diff
@@ -52,6 +47,9 @@ public override void Dispose()
         // Get the world size from the config, and use it to generate the correct-sized level
         protected virtual IEnumerator LoadWorld()
         {
+            // Defer a frame to allow worker flags to propagate.
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Assets/Fps/Scripts/SetupLogic/WorkerConnectorBase.cs]: https://improbableio.atlassian.net/browse/UTY-2116

File: workers/unity/Assets/Fps/Scripts/SetupLogic/ClientWorkerConnector.cs [hunk 7]
```diff
@@ -169,4 +176,60 @@ private void SendRequest()
...
+
+                if (textAsset == null)
+                {
+                    throw new MissingReferenceException("Unable to find DevAuthToken.txt in the Resources folder. " +
+                        "You can generate one via SpatialOS > Generate Dev Authentication Token.");
+                }
+
+                flow.DevAuthToken = textAsset.text;
+
+                return;
```
Comment (paulbalaji) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/ClientWorkerConnector.cs]: nit: redundant return

**Pull request merged**